Model/Details.php

```diff
@@ -0,0 +1,166 @@
+<?php
+
+/**
+ * Aligent Consulting
+ * Copyright (c) Aligent Consulting (https://www.aligent.com.au)
+ */
+
+declare(strict_types=1);
+
+namespace Aligent\AsyncEvents\Model;
+
+use Aligent\AsyncEvents\Api\AsyncEventRepositoryInterface;
+use Aligent\AsyncEvents\Model\ResourceModel\AsyncEventLog\Collection;
+use Aligent\AsyncEvents\Model\ResourceModel\AsyncEventLog\CollectionFactory as AsyncEventLogCollectionFactory;
+
+class Details
+{
+    /**
+     * @var array
+     */
+    private $traceCache = [];
+
+    /**
+     * @var Collection
+     */
+    private $collection;
+
+    /**
+     * @var AsyncEventRepositoryInterface
+     */
+    private $asyncEventRepository;
+
+    public function __construct(
+        AsyncEventLogCollectionFactory $collectionFactory,
+        AsyncEventRepositoryInterface  $asyncEventRepository
+    ) {
+        $this->collection = $collectionFactory->create();
+        $this->asyncEventRepository = $asyncEventRepository;
+    }
+
+    public function getDetails($uuid)
+    {
+        if (array_key_exists($uuid, $this->traceCache)) {
+            return $this->traceCache[$uuid];
+        }
+
+        $this->collection->addFilter('uuid', $uuid);
+
+        $traces = $this->collection->toArray();
+
+        if ($traces['totalRecords'] === 0) {
+            return [];
+        }
+
+        $asyncEventId = $this->collection->getFirstItem()->getData('subscription_id');
+        $asyncEvent = $this->asyncEventRepository->get($asyncEventId)->getData();
+
+        $this->traceCache[$uuid] = [
+            'traces' => $traces['items'],
+            'async_event' => $asyncEvent
+        ];
+
+        return $this->traceCache[$uuid];
+    }
+
+    public function getLogs($uuid)
+    {
+        $this->getDetails($uuid);
+
+        return $this->traceCache[$uuid]['traces'];
+    }
+
+    public function getStatus($uuid)
+    {
+        $this->getDetails($uuid);
+
+        $traces = $this->traceCache[$uuid]['traces'];
```

Comment:
You're relying on `getDetails` populating the array fully - otherwise you're going to get an undefined index error.
`getDetails` can return an empty array.

Should make it safe by defaulting to an empty array or similar:
`$traces = $this->traceCache[$uuid]['traces'] ?? [];`

Reply:
I'm guessing the only case that could happen is in a situation similar to manually editing the url with a key that does not exist in the db. If so, then will checking for in the controller and redirecting back with an error message solve pretty much a lot of the existence checking errors?